pkg/cmd/dev/recorderdriven_test.go

```diff
@@ -0,0 +1,203 @@
+// Copyright 2022 The Cockroach Authors.
+//
+// Use of this software is governed by the Business Source License
+// included in the file licenses/BSL.txt.
+//
+// As of the Change Date specified in that file, in accordance with
+// the Business Source License, use of this software will be governed
+// by the Apache License, Version 2.0, included in the file
+// licenses/APL.txt.
+
+package main
+
+import (
+	"bytes"
+	"flag"
+	"fmt"
+	"io"
+	"io/ioutil"
+	"log"
+	stdos "os"
+	stdexec "os/exec"
+	"strings"
+	"testing"
+
+	"github.com/cockroachdb/cockroach/pkg/cmd/dev/io/exec"
+	"github.com/cockroachdb/cockroach/pkg/cmd/dev/io/os"
+	"github.com/cockroachdb/cockroach/pkg/testutils"
+	"github.com/cockroachdb/cockroach/pkg/testutils/skip"
+	"github.com/cockroachdb/datadriven"
+	"github.com/irfansharif/recorder"
+	"github.com/stretchr/testify/require"
+)
+
+// TestRecorderDriven makes use of datadriven to record (if --rewrite is
+// specified) or play back (if --rewrite is omitted) all operations executed by
+// individual `dev` invocations. The testcases are defined under
+// testdata/recorderdriven/*; each test file corresponds to a captured recording
+// found in testdata/recorderdriven/*.rec.
+//
+// DataDriven divvies up these files as subtests, so individual "files" are
+// runnable through:
+//
+//  		dev test pkg/cmd/dev -f TestRecorderDriven/<fname>
+// 	OR  go test ./pkg/cmd/dev -run TestRecorderDriven/<fname>
+//
+// Recordings are used to mock out "system" behavior. When --rewrite is
+// specified, attempts to shell out to bazel or perform other OS operations
+// (like creating, removing, symlinking filepaths) are intercepted and system
+// responses are recorded for future playback. To update the test files with new
+// capture data, try:
+//
+// 		go test ./pkg/cmd/dev -run TestRecorderDriven/<fname> -rewrite
+//
+// NB: This test is worth contrasting to TestDataDriven, where all operations
+// are run in "dry-run" mode when --rewrite is specified. Here we'll actually
+// shell out (and take time proportional to running the actual commands). In
+// dry-run mode (TestDataDriven) all exec and os commands return successfully
+// with no error, with an empty response. This makes it suitable for testing
+// workflows that don't make use of external state to execute actions (like
+// reading the set of targets from a file for e.g., or hoisting files from a
+// sandbox by searching through the file system directly).
+//
+// TODO(irfansharif): When --rewrite-ing, because these tests shell out to the
+// actual host system, it makes it difficult to run under bazel/dev (currently
```

Comment:
Nit: It would be nice if `./dev test pkg/cmd/dev --rewrite` failed -- right now it passes silently but doesn't update any of the test files, which is confusing.

Reply:
Done. If used with -v it would log that the test was skipped, but probably this is fine too.